### DISCUSSION TITLE
Closing  connection from server side -

### DISCUSSION THREAD

**Discussion by rc4u** [strongswan/strongswan#1677] [Q&A] [answered]
I have a situation with an Android client connected to our VPN server.
After establishing a connection and surfing through the VPN tunneling, we stopped using the Android device.
When checking the logs, after 45 minutes from when the connection was established, the server tries to send to the client packets that are not getting a response from the client. My guess it happens due to the fact that the client app is running in the background.

Here are the logs from the server:
_May  1 12:17:46 10[IKE] <android|1398> establishing CHILD_SA android{1324} reqid 689
May  1 12:17:46 10[ENC] <android|1398> generating CREATE_CHILD_SA request 2 [ N(REKEY_SA) SA No TSi TSr ]
May  1 12:17:46 10[NET] <android|1398> sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:17:50 15[IKE] <android|1398> retransmit 1 of request with message ID 2
May  1 12:17:50 15[NET] <android|1398> sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:17:58 06[IKE] <android|1398> retransmit 2 of request with message ID 2
May  1 12:17:58 06[NET] <android|1398> sending packet: from X.X.X.X4[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:18:10 09[IKE] <android|1398> retransmit 3 of request with message ID 2
May  1 12:18:10 09[NET] <android|1398> sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:18:34 01[IKE] <android|1398> retransmit 4 of request with message ID 2
May  1 12:18:34 01[NET] <android|1398> sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:19:16 07[IKE] <android|1398> retransmit 5 of request with message ID 2
May  1 12:19:16 07[NET] <android|1398> sending packet: from X.X.X.X[4500] to Y.Y.Y.Y[16587] (352 bytes)
May  1 12:20:31 11[IKE] <android|1398> giving up after 5 retransmits
May  1 12:20:31 11[CFG] <android|1398> lease fec1::47 by '1-224030338' went offline
May  1 12:20:31 11[CFG] <android|1398> lease 10.2.0.71 by '1-224030338' went offline_

And the server drops the connection of the client. On the client side, it looks like the connection is still on, but when trying to surf, it failed.

The configuration in the server:
_conn android
      fragmentation=yes
      keyexchange=ikev2
      leftauth=pubkey
      left=%defaultroute
      leftsubnet=::/0,0.0.0.0/0
      leftfirewall=yes
      leftcert=serverCert.pem
      leftsendcert=always
      right=%any
      rightauth=eap-md5
      rightsourceip=10.2.0.0/16,fec1::0/16
      eap_identity=%identity
      auto=add
      reauth=no_

Is it possible to delay the server generating the new CHILD_SA, or disable it ?

When this happens, we cannot surf the Internet on the client side, but restarting the VPN connection will fix this. Is it possible to identify this state on the client side so we can activate the reconnect? The problem is when we perform a URL request from the client application. The request is not running through the VPN tunneling, so we cannot identify this state.

**Comment by tobiasbrunner** [comment]
> My guess it happens due to the fact that the client app is running in the background.

It should get woken up if a packet arrives, unless the device has networking completely disabled (or e.g. the client's IP address has changed).


> Is it possible to delay the server generating the new CHILD_SA, or disable it ?

You can change the rekeying settings or disable active rekeying by the server completely via `rekey=no`.

> Is it possible to identify this state on the client side so we can activate the reconnect?

If you are referring to our Android app, it has DPD disabled on purpose to not affect battery life if no actual traffic is exchanged and there may not be any connectivity for a while, which could trigger an unecessary recreation of the connection. So changing the settings on the server is recommended (to clean out abandoned SAs, you could configure DPD with an interval of several hours).

**Reply by rc4u** [reply]
> If you are referring to our Android app, it has DPD disabled on purpose to not affect battery life if no actual traffic is exchanged and there may not be any connectivity for a while, which could trigger an unecessary recreation of the connection. So changing the settings on the server is recommended (to clean out abandoned SAs, you could configure DPD with an interval of several hours).

By setting the dpddelay and dpdtimeout parameters for several hours?

**Reply by tobiasbrunner** [reply, marked as answer]
Yes, you could set `dpddelay` to a high value and `dpdaction=clear`. Note that `dpdtimeout` has no effect on IKEv2 connections.

**Reply by rc4u** [reply]
> > My guess it happens due to the fact that the client app is running in the background.
> 
> It should get woken up if a packet arrives, unless the device has networking completely disabled (or e.g. the client's IP address has changed).
> 
I checked this status more, and we see that the IP does not change, and the Application continues sending data to our servers (data from other services that are running in the application), but stopped responding to packages that the VPN server sends. Can you think of a way to identify the reason?

**Reply by tobiasbrunner** [reply]
No idea. It seems very weird.

**Reply by rc4u** [reply]
> You can change the rekeying settings or disable active rekeying by the server completely via `rekey=no`.

If I want to increase the time when the server will do the rekeying, what parameter can I set in the config ? currently, it happened after 45 minutes

**Reply by tobiasbrunner** [reply]
See [the docs](https://docs.strongswan.org/docs/5.9/config/rekeying.html) (see [here](https://wiki.strongswan.org/projects/strongswan/wiki/fromipsecconf) for help with translating the settings back to the legacy ipsec.conf format).

**Reply by rc4u** [reply]
> If you are referring to our Android app, it has DPD disabled on purpose to not affect battery life if no actual traffic is exchanged and there may not be any connectivity for a while, which could trigger an unecessary recreation of the connection. So changing the settings on the server is recommended (to clean out abandoned SAs, you could configure DPD with an interval of several hours).

is it possible to activate a DPD on your Android app ?

**Reply by tobiasbrunner** [reply]
Only by changing the source code [here](https://github.com/strongswan/strongswan/blob/69914a2dc58987752feb50bdf05c3d2d4baef05a/src/frontends/android/app/src/main/jni/libandroidbridge/backend/android_service.c#L789-L807).

**Reply by puresight2021** [reply]
I am still trying to overcome the situation where the client thinks it is connected, but the server has already dropped the connection.
One idea is that if I am able to send an HTTP request through the VPN tunnel, I will be able to identify this status. But when I run the request from the CLIENT application, it is going directly to the Internet and not through the VPN tunneling. Is it possible to run the request through the VPN tunneling?

**Reply by tobiasbrunner** [reply]
> I am still trying to overcome the situation where the client thinks it is connected, but the server has already dropped the connection.

You might want to change the server config to avoid that (different DPD and retransmission settings).

> One idea is that if I am able to send an HTTP request through the VPN tunnel, I will be able to identify this status. 

That might not always be because the server closed the IKE_SA. Could also just be a general connectivity issue on the client (maybe just a short interruption). Also, what would you do if you notice this?

> But when I run the request from the CLIENT application, it is going directly to the Internet and not through the VPN tunneling. Is it possible to run the request through the VPN tunneling?

If you're referring to an app that's based on our app code, then that's because it excludes itself from the VPN (in order to bypass the blocking and regular TUN devices so it can use DNS and fetch CRL/OCSP):

https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/frontends/android/app/src/main/java/org/strongswan/android/logic/CharonVpnService.java#L1124-L1137

**Reply by rc4u** [reply]
it looks like we can reproduce this issue more frequently, when we restart the IPSEC service. In most cases, the client application recognizes the disconnection and tries to connect again, but sometimes does not recognize the disconnection of the server and remains in connected mode

**Reply by tobiasbrunner** [reply]
If you terminate the daemon, for every established IKE_SA a single message is sent to delete it. That may or may not reach the client.

**Reply by rc4u** [reply]
exactly, this is why I want to find a way , from the client side to check if the server still there...
